### PR TITLE
Unmap zero page for Arm64

### PR DESCRIPTION
**Pull request by Weichen81**
As discussed in #296 , we have to unmap the zero page for NULL pointer.

Test Results:

**test_zeropage**
```
**** Solo5 standalone test_zeropage ****

Solo5: Trap: EL1 Synchronous Abort caught
Solo5: Dump registers:
         ESR    : 0x0000000096000047
         PC     : 0x0000000000102db8
         LR     : 0x0000000000102dac
         PSTATE : 0x00000000800003c5
         x00 ~ x03: 0x0000000000000001 0x0000000000000001 0x0000000000000008 0xfefefefefefefeff
         x04 ~ x07: 0x00000000001045f8 0x00000000001045f8 0x0000000000000001 0x000000001ffffc98
         x08 ~ x11: 0x0000000000105000 0x0000000000000020 0x0000000000000010 0x000000001ffffca0
         x12 ~ x15: 0x0000000000000002 0x0000000000000020 0x000000001ffffc98 0x0000000000000000
         x16 ~ x19: 0x0000000000000000 0x0000000000000000 0x0000000000104563 0x0000000000106000
         x20 ~ x23: 0x0000000000001000 0x0000000000000000 0x0000000000000000 0x0000000000000000
         x24 ~ x27: 0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000
         x28 ~ x29: 0x0000000000000000 0x000000001fffffc0
Solo5: ABORT: cpu_aarch64.c:73: PANIC

```

**test_ssp**
```
**** Solo5 standalone test_ssp ****

1234567890123456789012345678901234567890
Solo5: ABORT: Stack corruption detected

```

@mato

### CODE REVIEW

File: tenders/hvt/hvt_cpu_aarch64.c [hunk 1]
```diff
@@ -31,39 +31,107 @@
...
-    pud = (struct pud*)(va_addr + phys_pud);
+enum PGT_LEVEL {
+    L1_TABLE = 1, // PUD
+    L2_TABLE = 2, // PMD
+    L3_TABLE = 3, // PTE
+};
 
-    /* Fill PUD entries */
+static uint64_t aarch64_create_pagetable(enum PGT_LEVEL level,
+                        uint8_t *va_addr, uint64_t start, uint64_t size,
```
Comment (mato) [inline tenders/hvt/hvt_cpu_aarch64.c]: Could you please rename `va_addr` to just `mem`, to make it obvious that this is `hvt->mem`?

File: tenders/hvt/hvt_cpu_aarch64.c [hunk 2]
```diff
@@ -31,39 +31,107 @@
...
+                        uint64_t attrs, uint64_t attrs_io)
+{
+    uint32_t idx;
+    uint64_t *pentry;
+    uint64_t phys_table_addr, block_size, out_address;
+
+    if (level == L1_TABLE) {
+        /* The block size of each PUD entry is 1GB */
```
Comment (mato) [inline tenders/hvt/hvt_cpu_aarch64.c]: The comments here are unnecessary and repetitive. I'd prefer that you remove them entirely, or add a suitable generic comment at the start of the if / else block explaining what it does.

File: tenders/hvt/hvt_cpu_aarch64.c [hunk 3]
```diff
@@ -31,39 +31,107 @@
...
+        block_size = PMD_SIZE;
+        /* Locate the page of PMD to do this translation */
+        phys_table_addr = PGT_PAGE_ADDR(PGT_PMD_START + DIV_ROUND_UP(start, PUD_SIZE));
+    } else if (level == L3_TABLE) {
+        /* The block size of each PTE entry is 4KB */
+        block_size = PAGE_SIZE;
+        /* Locate the page of PTE to do this translation */
+        phys_table_addr = PGT_PAGE_ADDR(PGT_PTE_START + DIV_ROUND_UP(start, PMD_SIZE));
+    } else
+        err(1, "Unknown page table level: %d\n", level);
```
Comment (mato) [inline tenders/hvt/hvt_cpu_aarch64.c]: This should just be `assert(false)` or similar, if it fails it's a bug, not a "user reportable error".

File: tenders/hvt/hvt_cpu_aarch64.c [hunk 4]
```diff
@@ -31,39 +31,107 @@
...
 
-        size -= PUD_SIZE;
+        size -= block_size;
     }
 
-    return phys_pud;
+    return phys_table_addr;
+}
+
+static void aarch64_unmap_zero_page(uint8_t *va_addr)
```
Comment (mato) [inline tenders/hvt/hvt_cpu_aarch64.c]: As in other places, `va_addr` should just be `mem`.

File: tenders/hvt/hvt_cpu_aarch64.c [hunk 5]
```diff
@@ -73,7 +141,7 @@ static uint64_t aarch64_mapping_virt_to_phys(uint8_t *va_addr,
 void aarch64_setup_memory_mapping(uint8_t *va_addr, uint64_t ram_size,
```
Comment (mato) [inline tenders/hvt/hvt_cpu_aarch64.c]: Please rename `va_addr` to `mem` and `uint64_t ram_size` to `size_t mem_size`.

**Comment by Weichen81** [comment]
@mato Thanks for your comments. I can address them, but it would cost some times, I think it would not be done before this weekend or next Monday. Is it OK?

**Comment by mato** [comment]
@weichen81 No hurry, take your time.

**Comment by Weichen81** [comment]
New PR create https://github.com/Solo5/solo5/pull/300

**Pull request closed without merging**